svc/GridAutosizeService.js

```diff
@@ -0,0 +1,285 @@
+/*
+ * This file belongs to Hoist, an application development toolkit
+ * developed by Extremely Heavy Industries (www.xh.io | info@xh.io)
+ *
+ * Copyright © 2020 Extremely Heavy Industries Inc.
+ */
+
+import {HoistService} from '@xh/hoist/core';
+import {stripTags} from '@xh/hoist/utils/js';
+import {isFunction, isNil, isFinite, sortBy, groupBy, map, reduce, min, max} from 'lodash';
+
+/**
+ * Calculates the column width required to display all values in a column.
+ *
+ * Unlike the native ag-Grid autosizing, this service will compute a size based on all
+ * data in the grid, including off-screen rows and columns, and collapsed rows.
+ *
+ * In order to do this efficiently, this service uses heuristics and that generally assumes each
+ * column consists of similarly formatted strings.  In particular, it cannot make the computation
+ * when a react based elementRenderer has been specified for a column. In this case, no width will
+ * be computed, and the column will be ignored by this service.
+ *
+ * @see Column.autoSizeOptions for options to control this behaviour
+ */
+@HoistService
+export class GridAutosizeService {
+
+    _canvasContext;
+    _headerEl;
+    _cellEl;
+
+    /**
+     * Calculates the required column widths for a GridModel. Returns an array of the
+     * form [{colId, width}] suitable for consumption by GridModel.applyColumnStateChanges().
+     * Typically called via `GridModel.autoSizeColumns()`.
+     *
+     * @param {GridModel} gridModel - GridModel to autosize.
+     * @param {String[]} colIds - array of columns in model to compute sizing for.
+     */
+    autoSizeColumns(gridModel, colIds) {
+        const ret = [],
+            {store} = gridModel;
+
+        // Get filtered set of records
+        let records = [];
+        if (gridModel.agApi?.isAnyFilterPresent()) {
+            gridModel.agApi.forEachNodeAfterFilter(node => {
+                const record = store.getById(node.data?.id);
+                if (record) records.push(record);
+            });
+        } else {
+            records = [...store.records];
+        }
+
+        if (store.summaryRecord) {
+            records.push(store.summaryRecord);
+        }
+
+        for (const colId of colIds) {
+            const width = this.autoSizeColumn(gridModel, records, colId);
+            if (isFinite(width)) ret.push({colId, width});
+        }
+
+        return ret;
+    }
+
+    //------------------
+    // Implementation
+    //------------------
+    autoSizeColumn(gridModel, records, colId) {
+        const column = gridModel.findColumn(gridModel.columns, colId);
```

Comment:
This was always a weird signature for `gridModel.findColumn` - not sure we can change it now, but it would have been nice if `colId` were the first param and there was a second optional param that let you specify a subset of columns to search (but defaulted to all the gridModel's columns).